pulsar-broker/src/main/java/org/apache/pulsar/PulsarClusterMetadataTeardown.java

```diff
@@ -107,5 +135,77 @@ public static void deleteZkNodeRecursively(ZooKeeper zooKeeper, String path) thr
         }
     }
 
+    private static List<String> getChildren(ZooKeeper zooKeeper, String path) {
+        try {
+            return zooKeeper.getChildren(path, null);
+        } catch (InterruptedException | KeeperException e) {
+            if (e instanceof KeeperException.NoNodeException) {
+                return new ArrayList<>();
+            }
+            log.error("Failed to get children of {}: {}", path, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static byte[] getData(ZooKeeper zooKeeper, String path) {
+        try {
+            return zooKeeper.getData(path, null, null);
+        } catch (KeeperException | InterruptedException e) {
+            log.error("Failed to get data from {}: {}", path, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void deleteLedger(BookKeeper bookKeeper, long ledgerId) {
+        try {
+            bookKeeper.deleteLedger(ledgerId);
+            if (log.isDebugEnabled()) {
+                log.debug("Delete ledger id: {}", ledgerId);
+            }
+        } catch (InterruptedException | BKException e) {
+            log.error("Failed to delete ledger {}: {}", ledgerId, e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static void deleteManagedLedgers(ZooKeeper zooKeeper, ManagedLedgerFactory managedLedgerFactory) {
+        final String managedLedgersRoot = "/managed-ledgers";
+        getChildren(zooKeeper, managedLedgersRoot).forEach(tenant -> {
+            final String tenantRoot = managedLedgersRoot + "/" + tenant;
+            getChildren(zooKeeper, tenantRoot).forEach(namespace -> {
+                final String namespaceRoot = String.join("/", tenantRoot, namespace, "persistent");
+                getChildren(zooKeeper, namespaceRoot).forEach(topic -> {
+                    final TopicName topicName = TopicName.get(String.join("/", tenant, namespace, topic));
+                    try {
+                        managedLedgerFactory.delete(topicName.getPersistenceNamingEncoding());
+                    } catch (InterruptedException | ManagedLedgerException e) {
+                        log.error("Failed to delete ledgers of {}: {}", topicName, e);
+                        throw new RuntimeException(e);
+                    }
+                });
+            });
+        });
+    }
+
+    private static void deleteSchemaLedgers(ZooKeeper zooKeeper, BookKeeper bookKeeper) {
+        final String schemaLedgersRoot = "/schemas";
```

Comment:
Probably all of this direct access to Zookeeper is to be moved inside ManagedLedgerFactory.

Wdyt @sijie?

Reply:
Schema's ledgers are managed by `BookkeeperSchemaStorage`, but it's coupled with `PulsarService` currently because it's a part of `pulsar-broker`, not an independent module like `managed-ledger`. So it uses `PulsarService`'s BK client now. The change needs some refactor of schema storage.

Reply:
However, since `BookkeeperSchemaStorage` only uses `PulsarService` for creating a `BookKeeper` interface, the refactor may not need too much changes. I'll try to refactor it later.

Reply:
After looking into code again, I think it's not proper to use `BookkeeperSchemaStorage` here. Because `BookKeeper` will be initialized until `start()` method is invoked. If a constructor which accepts a `BookKeeper` instance was added, the `BookkeeperSchemaStorage` shouldn't invoke `start()` and some other methods. On the other hand, the current implementation of `BookkeeperSchemaStorage#delete` doesn't access ZooKeeper. So at this moment, we need to access to ZooKeeper directly.